src/providers/hana/qgshanaconnection.cpp

```diff
@@ -0,0 +1,621 @@
+  try
+  {
+    StatementRef stmt = mConnection->createStatement();
+    ResultSetRef rsSchemas = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+    while ( rsSchemas->next() )
+    {
+      QgsHanaSchemaProperty schema;
+      schema.name = QgsHanaUtils::toQString( rsSchemas->getNString( 1 ) );
+      schema.owner = QgsHanaUtils::toQString( rsSchemas->getNString( 2 ) );
+      list << schema;
+    }
+    rsSchemas->close();
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+  return list;
+}
+
+int QgsHanaConnection::getLayerSRID( const QgsHanaLayerProperty &layerProperty )
+{
+  if ( layerProperty.geometryColName.isEmpty() )
+    return -1;
+
+  int ret = -1;
+
+  try
+  {
+    StatementRef stmt = mConnection->createStatement();
+
+    if ( !layerProperty.isView )
+    {
+      QString sql = QStringLiteral( "SELECT SRS_ID FROM SYS.ST_GEOMETRY_COLUMNS "
+                                    "WHERE SCHEMA_NAME = '%1' AND TABLE_NAME = '%2' AND COLUMN_NAME = '%3'" )
+                    .arg( layerProperty.schemaName, layerProperty.tableName, layerProperty.geometryColName );
+      ResultSetRef rsSrid = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+      ret = rsSrid->next() ? *rsSrid->getInt( 1 ) : -1;
+      rsSrid->close();
+    }
+    else
+    {
+      QString sql = QStringLiteral( "SELECT %1.ST_SRID() FROM %2.%3 WHERE %1 IS NOT NULL LIMIT %4" )
+                    .arg( QgsHanaUtils::quotedIdentifier( layerProperty.geometryColName ),
+                          QgsHanaUtils::quotedIdentifier( layerProperty.schemaName ),
+                          QgsHanaUtils::quotedIdentifier( layerProperty.tableName ),
+                          QString::number( GEOM_TYPE_SELECT_LIMIT ) );
+      ResultSetRef rsSrid = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+      int srid = -1, prevSrid = -1;
+      while ( rsSrid->next() )
+      {
+        srid = *rsSrid->getInt( 1 );
+        if ( prevSrid != -1 && srid != prevSrid )
+        {
+          srid = -1;
+          break;
+        }
+        prevSrid = srid;
+      }
+
+      ret = srid;
+      rsSrid->close();
+    }
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  return ret;
+}
+
+QStringList QgsHanaConnection::getLayerPrimaryeKeys( const QgsHanaLayerProperty &layerProperty )
+{
+  QStringList ret;
+
+  try
+  {
+    DatabaseMetaDataRef dbmd = mConnection->getDatabaseMetaData();
+    ResultSetRef rsPrimaryKeys = dbmd->getPrimaryKeys( nullptr,
+                                 layerProperty.schemaName.toStdString().c_str(),
+                                 layerProperty.tableName.toStdString().c_str() );
+
+    while ( rsPrimaryKeys->next() )
+    {
+      auto keyName = rsPrimaryKeys->getNString( 4 );
+      if ( !keyName.isNull() )
```

Comment:
I don't think `keyName` can ever be `NULL`. That would not make any sense.

Reply:
Done.